File: Arrays/Easy/Removes Duplicates From Sorted Array/RemoveDuplicatesFromSortedArray.cpp
#include <bits/stdc++.h>
using namespace std;

/*------------------------------ Solution Class -----------------------------*/
class Solution
{
public:
    // Time Complexity:- O(n), n is the size of vector.
    // Space Complexity:- O(1)
    int removeDuplicates(vector<int> &nums)
    {
        int i = 0;
        for (int j = 1; j < nums.size(); j++)
        {
            if (nums[i] != nums[j])
            {
                i++;
                nums[i] = nums[j];
            }
        }
        return i + 1;
    }

    /* Method to print vector. */
    void print(vector<int> arr, int k)
    {
        for (int i = 0; i < k; i++)
        {
            cout << arr[i] << " ";
        }
        cout << endl;
    }
};
/*-------------------------- Solution Class Ends Here. ------------------------- */
int main()
{
    vector<int> arr = {0, 0, 1, 1, 1, 2, 2, 3, 3, 4};

    Solution s;
    int newSize = s.removeDuplicates(arr);
    s.print(arr, newSize);
}

File: 1. Arrays/Easy/Removes Duplicates From Sorted Array/RemoveDuplicatesFromSortedArray.java


public class RemoveDuplicatesFromSortedArray {
    // Time Complexity:- O(n), n is the size of vector.
    // Space Complexity:- O(1)
    static int removeDuplicates(int nums[])
    {
        int n = nums.length;

        int i = 0;
        for (int j = 1; j < n; j++) {
            if(nums[i] != nums[j])
            {
                // move "i" one step.
                i++;

                // swap nums[i], nums[j].
                nums[i] = nums[i] ^ nums[j];
                nums[j] = nums[i] ^ nums[j];
                nums[i] = nums[i] ^ nums[j];

            }
        }

        return (i+1);

    }
    public static void main(String[] args) {
        int arr[] = {0, 0, 1, 1, 1, 2, 2, 3, 3, 4};

        int idx = removeDuplicates(arr);
        for (int i = 0; i < idx; i++) {
            System.out.print(arr[i] + " ");
        }
    }
}
